src/geometry.rs

```diff
@@ -0,0 +1,180 @@
+// }
+
+// Чтобы правильно рассчитывать освещенность в точке, нужно считать нормаль в этой точке.
+// Кроме того, чтобы поддерживать текстуры необходимо рассчитывать барацентрические
+// координаты точки в треугольнике, а для каждой вершины треугольника помнить координаты
+// этой верщины в текстуре. Кроме того, для улучшения качества нужно помнить нормали в
+// каждой вершине, и рассчитывать нормаль в точке как взвешенную сумму нормалей в
+// вершинах, с весами, равными барацентрическим координатам.
+
+// struct Sphere {...}
+// Хотим поддержку шариков ведь, так?)
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn zero_distance_between_point_and_itself() {
+        let point = Point(1., 4., 5.);
+        assert_eq!(point.distance(&point), 0.0);
+    }
+
+    #[test]
+    fn distance_is_positive() {
+        let point = Point(1., 4., 5.);
+        let other_point = Point(2., 1., -4.);
+        assert!(point.distance(&other_point) > 0.0);
+    }
+
+    #[test]
+    fn distance_is_symmetric() {
+        let point = Point(1., 4., 5.);
+        let other_point = Point(2., 1., -4.);
+        assert_eq!(point.distance(&other_point), other_point.distance(&point));
+    }
+
+    #[test]
+    fn triangle_inequality() {
+        let a_point = Point(1., 4., 5.);
+        let b_point = Point(2., 1., -4.);
+        let c_point = Point(2., -1., 2.);
+        assert!(
+            a_point.distance(&b_point) + b_point.distance(&c_point) >= a_point.distance(&c_point)
+        );
+    }
+
+    #[test]
+    fn distance_gives_right_answer() {
+        let a_point = Point(1., 1., 2.);
+        let b_point = Point(1., 1., 0.);
+        assert_eq!(a_point.distance(&b_point), 2.);
+        let a_point = Point(1., 1., 2.);
+        let b_point = Point(1., 4., 2.);
+        assert_eq!(a_point.distance(&b_point), 3.);
+        let a_point = Point(7., 1., 0.);
+        let b_point = Point(1., 1., 0.);
+        assert_eq!(a_point.distance(&b_point), 6.);
+    }
+
+    #[test]
+    fn shift_by_zero_vector() {
+        let point = Point(1., 5., 6.);
+        let vector = Vector(0., 0., 0.);
+        assert_eq!(point.0, point.shift_by(&vector).0);
+        assert_eq!(point.1, point.shift_by(&vector).1);
+        assert_eq!(point.2, point.shift_by(&vector).2);
+        assert_eq!(0., vector.norm());
+    }
+
+    #[test]
+    fn normalizing_vector() {
+        let vector = Vector(1., 3., 7.);
+        assert_eq!(1., vector.normalize().norm());
+    }
+
+    #[test]
+    #[should_panic]
+    fn normalizing_zero_vector() {
+        let zero_vector = Vector(0., 0., 0.);
+        zero_vector.normalize();
+    }
+
+    #[test]
+    fn vector_product_for_standard_basis() {
+        let a_vector = Vector(1., 0., 0.);
+        let b_vector = Vector(0., 1., 0.);
+        let product = a_vector.vector_product(&b_vector);
+        assert_eq!(0., product.0);
+        assert_eq!(0., product.1);
+        assert_eq!(1., product.2);
```

Comment:
Is it possible to write custom equality check for vector to avoid code duplication?